readthedocs/projects/tasks.py

```diff
@@ -87,6 +89,94 @@
 log = logging.getLogger(__name__)
 
 
+class CachedEnvironmentMixin:
+
+    """Mixin that pull/push cached environment to storage."""
+
+    def pull_cached_environment(self):
+        if not self.project.has_feature(feature_id=Feature.CACHED_ENVIRONMENT):
+            return
+
+        storage = get_storage_class(settings.RTD_BUILD_ENVIRONMENT_STORAGE)()
+        filename = self.version.get_storage_environment_cache_path()
+
+        msg = 'Checking for cached environment'
+        log.debug(
+            LOG_TEMPLATE,
+            {
+                'project': self.project.slug,
+                'version': self.version.slug,
+                'msg': msg,
+            }
+        )
+        if storage.exists(filename):
+            msg = 'Pulling down cached environment from storage'
+            log.info(
+                LOG_TEMPLATE,
+                {
+                    'project': self.project.slug,
+                    'version': self.version.slug,
+                    'msg': msg,
+                }
+            )
+            tmp_filename = tempfile.mkstemp(suffix='.tar')
+            remote_fd = storage.open(filename, mode='rb')
+            with open(tmp_filename, mode='wb') as local_fd:
+                local_fd.write(remote_fd.read())
+
+            with tarfile.open(tmp_filename) as tar:
+                tar.extractall(self.version.project.doc_path)
```

Comment:
I wonder if it's worth abstracting the archive format here. It looks like `shutil` has an abstraction over it. I imagine we'll want to play with `tar` and `gzip` at least, because I imagine bandwidth and CPU will be a tradeoff here that we want to be able to tweak.

https://docs.python.org/3/library/shutil.html#archiving-operations

Reply:
Hrm... Interesting! I didn't know about `shutil.make_archive`. I will take a deeper look.

Reply:
Having an abstraction is a good idea to be able to play here. Although, it seems that `shutil.make_archive` does not gives us too much ability for this since it's not too customizable: we can't change lower level attributes.

Besides, we would need Python 3.8 because since that version they are using modern PAX (``--xattrs`` equivalent):

> Changed in version 3.8: The modern pax (POSIX.1-2001) format is now used instead of the legacy GNU format for archives created with format="tar".

On the other hand, `tarfile` allows us to play with `gzip`, `bz2` and `lzma` and with compression level as well.